src/DesktopAnalytics/Analytics.cs

```diff
@@ -369,20 +398,32 @@ private void ReportIpAddressOfThisMachineAsync()
 				try
 				{
 					Uri uri;
-					Uri.TryCreate(UrlThatReturnsExternalIpAddress, UriKind.Absolute, out uri);
+					bool json = string.IsNullOrEmpty(UrlThatReturnsExternalIpAddress);
+					Uri.TryCreate(json ? UrlThatReturnsGeolocationJson : UrlThatReturnsExternalIpAddress, UriKind.Absolute, out uri);
 					client.DownloadDataCompleted += (object sender, DownloadDataCompletedEventArgs e) =>
 					{
 						try
 						{
-							var externalIpAddress = System.Text.Encoding.UTF8.GetString(e.Result).Trim();
-							Debug.WriteLine(String.Format("DesktopAnalytics: external ip = {0}", externalIpAddress));
-							_options.Context.Add("ip", externalIpAddress);
-							_propertiesThatGoWithEveryEvent.Add("ip", externalIpAddress);
+							var result = System.Text.Encoding.UTF8.GetString(e.Result).Trim();
+							if (json)
+							{
+								Debug.WriteLine(String.Format("DesktopAnalytics: geolocation JSON data = {0}", result));
+								var j = new JTokenReader(JToken.Parse(result));
+								AddGeolocationProperty(j, "city");
+								AddGeolocationProperty(j, "country", "countryCode");
+								AddGeolocationProperty(j, "regionName", "region");
+							}
+							else
+							{
+								Debug.WriteLine(String.Format("DesktopAnalytics: external ip = {0}", result));
+								_options.Context.Add("ip", result);
+								_propertiesThatGoWithEveryEvent.Add("ip", result);
+							}
 						}
 						catch (Exception)
 						{
 							// we get here when the user isn't online, or anything else prevents us from
-							// getting their ip. Still worth reporting the launch in the later case.
+							// getting their ip. Still worth reporting the launch in the latter case.
```

Comment:
1) Getting there *location*"

2) I don't understand the original code... if we get the info we want, we add a new piece of info to the properties that go with ever event. Cool. But if we don't get the info, then we call `TrackWith...("Launch")`?

Reply:
I guess the point is that as long as they are on-line, we can still note that the program was launched (so we can count it) even if we couldn't figure out anything about the location. This is probably a pretty unlikely scenario, but if the URL(s) for getting that info disappear or are blocked by a particular country or ISP, we can at least know the program is being used somewhere.